Properly marks k8s stream complete on error exit

tcsc opened the pull request:
Addresses #5624

When a k8s stream exits it emits events that mark the session recording
as complete. Prior to this patch, when the `exec` handler exited before
emitting these "session complete" events due to an error, leaving the
recordings orphaned.

This patch wraps the stream cleanup in a `defer`ed cleanup handler
that marks the stream as complete in any exit mode.

klizhentas commented on lib/kube/proxy/forwarder.go:
```diff
@@ -804,7 +804,7 @@ func (f *Forwarder) exec(ctx *authContext, w http.ResponseWriter, req *http.Requ
 	defer proxy.Close()
 
-	f.log.Debugf("Created streams, getting executor.")
+	f.log.Debugf("Getting executor")
```
I suggest to remove this log entry, I assume it will be very verbose as it's called every time

klizhentas commented on lib/kube/proxy/forwarder.go:
```diff
@@ -813,6 +813,8 @@ func (f *Forwarder) exec(ctx *authContext, w http.ResponseWriter, req *http.Requ
 	}
 	streamOptions := proxy.options()
 
+	f.log.Debugf("Binding streams")
```
Same here

klizhentas commented on lib/kube/proxy/forwarder.go:
```diff
@@ -834,127 +836,137 @@ func (f *Forwarder) exec(ctx *authContext, w http.ResponseWriter, req *http.Requ
+	// whether it exits successfully, or with an error.
+	// NOTE that this cleanup handler MAY MODIFY the returned error value.
+	defer func() {
+		f.log.Debug("Cleaning up after stream exit")
```
Same here

klizhentas commented on lib/kube/proxy/forwarder.go:
```diff
@@ -834,127 +836,137 @@ func (f *Forwarder) exec(ctx *authContext, w http.ResponseWriter, req *http.Requ
 	}
 
-	f.log.Debugf("Exited successfully.")
+	f.log.Debug("Stream finished without error")
```
I would remove this one too.

klizhentas commented on lib/kube/proxy/forwarder.go:
```diff
@@ -834,127 +836,137 @@ func (f *Forwarder) exec(ctx *authContext, w http.ResponseWriter, req *http.Requ
 		}
+	}()
+
+	f.log.Debug("Running stream")
```
Same here - if it's happy path running normally, I would not add debug logs

klizhentas commented on lib/kube/proxy/forwarder.go:
```diff
@@ -834,127 +836,137 @@ func (f *Forwarder) exec(ctx *authContext, w http.ResponseWriter, req *http.Requ
-		f.log.WithError(err).Warning("Failed to send status. Exec command was aborted by client.")
-		return nil, trace.Wrap(err)
-	}
+	// Defer a clanup handler that will mark the stream as complete on exit, regardless of
```
`cleanup`

tcsc committed:
Properly marks k8s stream complete on error exit

Addresses #5624

When a k8s stream exits it emits events that mark the session recording
as complete. Prior to this patch, the `exec` handler exited before
emitting these "session complete" events, leaving the recordings
orphaned.

This patch wraps the stream cleanup in a `defer`ed cleanup handler
that marks the stream as complete in any exit mode.
